### PR TITLE
Global UUID registry, cleanup, simplification

### PR DESCRIPTION
**Pull request by andyross**
This is a situation where scratching an itch (trying to get a clean mapping of component "names" I could track with some topology tooling I was experimenting with) turned into a big bleeding wound.  But the healing is probably worth it.

Basically SOF's UUID hygiene had broken down in a bunch of places, and the cleanup led to a global registry that should be simpler to maintain in the long term.  There's lots of code motion here but no signficant behavior changes, just a new and cleaner API.  Please read the commit messages carefully, but the executive summary is:

* No more typing UUIDs into C code.  All IDs are defined with unique names via `SOF_DEFINE_REG_UUID(name);`

* A handful of spots where we had colliding names and (gulp) IDs have been corrected.

* IDs are specified in a minimally complicated text file at the top of the source tree.  The build time C header generation automatically validates this file so it will stay clean.

* UUID string names (for IDs referenced by the build) are available at runtime in Zephyr builds via the pre-existing sof_uuid_entry struct, and you can enumerate over them via the Zephyr iterable sections API.  XTOS builds remain compatible, but don't get the new feature.  (This does have an image size cost of ~1-2k in existing firmware builds.)

* There remain a few spots where separate components share UUIDs.  These have been left in place for compatibility, though obviously they can't (and haven't been) be built into a single image.  The names of those IDs have been unified, however.

Longer term the hope is that this can be unified with things like topology and manifest generation, both to prevent errors due to miscopied IDs and to simplify tooling by giving users the ability to use a plaintext name instead.

~~Note also that this requires a Zephyr PR to build successfully, which corrects some mistakes with "snippets" in the Intel and NXP linker scripts.  Builds without them drop the runtime UUID structs into invalid or orphan addresses (which are linker warnings, but not errors) and will cause crashes at runtime:  https://github.com/zephyrproject-rtos/zephyr/pull/75055~~

Actually all that's needed is the ROM_SECTIONS PR that already merged upstream, my fixup crossed, and there's no reason not to use the variant that merged.  The next Zephyr uprev (e.g. to the 3.7.0 LTS version which is mere hours from shipping) should catch this automatically

### CODE REVIEW

**Comment by andyross** [comment]
Heh, look at all the reviewers.  Tree-wide changes add basically everyone, I guess.  FWIW I don't expect this to progress that quickly, it'll likely take a while to duplicate and resolve all the CI failures I haven't run locally.  And as mentioned there's a Zephyr PR that needs to merge first anyway.  Please review though.

File: src/include/sof/lib/uuid.h
Comment (andyross) [inline src/include/sof/lib/uuid.h]: Oops, squashme patch wasn't and got left in, will fix

File: src/include/sof/lib/uuid.h
Comment (cujomalainey) [inline src/include/sof/lib/uuid.h:14]: it would probably be better to abstract this into an arch macro

Comment (andyross) [reply src/include/sof/lib/uuid.h:14]: That's largely what this is.  "`__ZEPHYR__`" is set in all C code compiled under a Zephyr app (via a setting in the top-level Zephyr CMakeLists.txt), it's used elsewhere in SOF as a tag for Zephyr-vs-legacy.  The usage here is an attempt to be minimal, basically each of Zephyr and XTOS define three small(ish) macros which are then used to implement the main UUID API.

Comment (cujomalainey) [reply src/include/sof/lib/uuid.h:14]: Yes, we have just been trying to abstract this into the "rtos" headers, this would be a "nice to have" but ill let others make the call on how much we want to be pristine here

Comment (andyross) [reply src/include/sof/lib/uuid.h:14]: Ah, I get it.  You mean move the helper macros down into `{zephyr,xtos}/include/rtos/uuid.h` files and then just "`#include <rtos/uuid.h>`" here to get them?

That could be done, sure.  I guess my sense is that's more work than worth it for three macros in seven lines over two variants, but would be easy enough.

Comment (cujomalainey) [reply src/include/sof/lib/uuid.h:14]: yep exactly :)

File: src/audio/dai-legacy.c
Comment (lyakh) [inline src/audio/dai-legacy.c:44]: is this part of the unification needed for some automation in one of the following commits? If not - maybe we could keep different names, makes grepping the code easier

File: scripts/gen-uuid-reg.py
Comment (lyakh) [inline scripts/gen-uuid-reg.py:30]: bytes?

File: uuid-registry.txt
Comment (lyakh) [inline uuid-registry.txt:158]: I was thinking about sorting it by UUIDs - then of course any duplication attempts would be immediately obvious, but that is checked by your scripting anyway, so probably sorting by name is better indeed

File: src/audio/eq_iir/eq_iir.c
Comment (lyakh) [inline src/audio/eq_iir/eq_iir.c:40]: can we also automate LLEXT UUUID uses? https://github.com/thesofproject/sof/blob/d51fa68079aed2dcb8f27f6e2eaffe15c79e7850/src/audio/eq_iir/eq_iir.c#L266

Comment (andyross) [reply src/audio/eq_iir/eq_iir.c:40]: Didn't notice this one.  Will do.

File: scripts/gen-uuid-reg.py
Comment (andyross) [reply scripts/gen-uuid-reg.py:30]: "bytes()" is a python builtin, alas.

File: src/audio/dai-legacy.c
Comment (andyross) [reply src/audio/dai-legacy.c:44]: Actually this is sort of a fakeout.  The names in-tree are both "dai", and they were renamed by an early patch in this series to disambiguate them.  But then it turns out that the **value** of the UUID is the same in both cases, which is unfortunate but technically unresolvable (I mean... not really, but I'm trying very hard not to have to renumber these things and being really clear where I do).  So this moves back to the original usage where the two drivers share the same UUID, which is OK in this case as they can't be built into the same image anyway.

I should fold them together to reduce the churn, though that's a mildly hard rebase.

File: scripts/xtensa-build-zephyr.py
Comment (andyross) [reply scripts/xtensa-build-zephyr.py:901]: Honestly I'm hoping someone chimes in to clarify this.

AFAICT sof-logger, smex, .ldc files and old-style trace context usage are all degenerate in Zephyr builds.  The window that used to store the dictionary logging output has been repurposed to transmit plaintext Zephyr logs, and almost nothing is actually emitting the trace output anyway as all the {tr,comp}_{err,warn,info}() has been redirected to Zephyr LOG_xxx() macros which ignore the context struct.

Basically this is me ripping the bandaid off to see what breaks.  For sure, after this PR merges none of that stuff will work anymore as the addresses of the sof_uuid structs won't appear within the static_uuids section and sof-logger won't see any ldc entries.

If we absolutely still need this, let me know and I'll see what needs to be ported.

Comment (kv2019i) [reply scripts/xtensa-build-zephyr.py:901]: @andyross @lyakh Let's have some from @thesofproject/nxp check this. In the transition to Zephyr, we did use sof-logger early with Zephyr so the smex/ldc/soflogger stuff was also used with Zephyr, but indeed for Intel platforms we've now moved on to Zephyr logging subsystem completely. But has NXP made the move? There is certainly a maintainance cost in having these around (see https://github.com/thesofproject/sof/pull/8822 ), so in general I'm happy to drop these.

PS The UUID rework looks great. Once we have the v2.10 release out, I'll do a more proper review.

Comment (cujomalainey) [reply scripts/xtensa-build-zephyr.py:901]: @thesofproject/mediatek should also check to see if they still depend on this

Comment (andyross) [reply scripts/xtensa-build-zephyr.py:901]: FWIW: mt8xxx builds are fine in-tree as non-Zephyr builds are unchanged and will continue to work.

The question here is whether or not any Zephyr configurations are still in the deprecated/only-partially-working trace contexts.   And I'm pretty sure[1] the answer is no, because if you tried you'd discover that it doesn't actually have all the info.  When __ZEPHYR__==1, all the tr_info()/comp_info() logging et. al. goes to a Zephyr LOG_INFO() macro, 100% of the time.  And the implementation of that ignores the trace context that would have been used earlier to associate the log message with a UUID[2].  So the only output you'll get out of it are a handful of spots that are still going to the trace context (buffer and dai logging, in particular, are still using trace-based macros which land in stubs in a Zephyr build, though AFAICT those don't appear on the Zephyr output and no one seems to care!).

Anyway, the point being: this PR moves the UUIDs to a different location in the binary (an iterable array and not a special `.static_uuids` section) with a different layout (the `struct sof_uuid_entry` that includes the string name in addition to the bare UUID).  So it's 100% going to break any existing users.  I just don't think there are any.

[1] I mean, actually I'm more than pretty sure.  But software is complicated and edge cases abound, thus the questions.

[2] Zephyr logging is plaintext, ignores the trace context and UUID, and doesn't require .ldc files nor smex/sof_logger.  There actually is a Zephyr dictionary logging feature @dcpleung did long ago, but it's not enabled for SOF and to be honest I don't know much about how its tooling works.

Comment (cujomalainey) [reply scripts/xtensa-build-zephyr.py:901]: Oh i totally misread this as C code, not python, then yes ignore my previous comment since this is zephyr specific

Comment (marc-hb) [reply scripts/xtensa-build-zephyr.py:901]: Unfortunately we still have old and very complex build and CI dependencies here because (as usual) the logging transitions to Zephyr and IPC4 were after-thoughts and rushed on the validation side. Some in Intel CI are not even open-source, which means you couldn't fix them even if you volunteered to.

So unless this literally **stops building**, I highly recommend you don't touch this if you want the rest merged in a reasonable time. Feel free to print additional warnings.

**Review by marc-hb — COMMENTED** [review]
I'm switching to a draft for now for the simple reason that it seems to fail most CI checks.

Speaking of CI, here's my usual broken record:

https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

It's great to share (as a draft!) the entire thing, this clearly shows the direction and rationale. But as far as review and CI are concerned, please break this down into smaller PRs that can be reviewed, tested, debugged and merged _faster_.

**Comment by marc-hb** [comment]
> https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#defining-smaller-prs

In case any evidence was still required, here's the example of the week where a very small, totally unrelated commit had some sort of "Butterfly effect" and triggered **DSP panics (!) all across the MTL board**:

- https://github.com/thesofproject/sof/issues/9268

Luckily enough, reproduction for this one is deterministic - it's usually not the case.

It's not the first time and certainly the last, so: no giant PR. Especially not when it's _already_ broken down in 13 separate commits.

**Comment by andyross** [comment]
> So unless this literally stops building, I highly recommend you don't touch this if you want the rest merged in a reasonable time. Feel free to print additional warnings.

I'm happy to wait for merge, but IMHO that's exactly backwards, and largely how we got where we are with subsystems like this (in this case: Zephyr builds literally carrying around artifacts for a logging system that doesn't work anymore).  We need to be willing to "finish the job" with big changes, even (especially!) when we don't necessarily know what's going to break if we touch it.  If we don't know now, we're **even less likely** to understand it after further modification.  And so we'll be even more resistant to change the next time someone want to update it, and so on.

You fix "stuff breaks with changes" by maximizing software robustness and simplifying design, not by refusing to change stuff.  (Also: that MTL power_down() failure is deep hardware voodoo and not at all just a software mistake.  That stuff just happens and you can't prevent it with process.)

As far as size, I'll split up what I can.  But these aren't unrelated changes.  If you read through, they're steps along the way to make clearer.  You could, in theory, stop at any one of them.  But you wouldn't want to.  And many of them are treewide, large API changes that would be harder to review sequentially without context.

**Comment by marc-hb** [comment]
> You fix "stuff breaks with changes" by maximizing software robustness and simplifying design, not by refusing to change stuff. 

Very happy you're volunteering to fix all the tests and infrastructure. We need more people like you!

**Comment by marc-hb** [comment]
> You could, in theory, stop at any one of them. But you wouldn't want to.

Yes you do for simple testing purposes. It's called "Continuous" Integration.

> And many of them are treewide, large API changes that would be harder to review sequentially without context.

There are multiple levels of reviews. There is (at least):
- the great picture
- the tedious, line-by-line bug catching.

This giant PR successfully provides the first one. Smaller PRs solve the second one.

File: scripts/xtensa-build-zephyr.py
Comment (kv2019i) [reply scripts/xtensa-build-zephyr.py:901]: @andyross As for Zephyr logging being plaintext, we have enabled both Zephyr dictionary and mipi syst and even have an overlay in upstream SOF -> work/sof/app/logging_mipisystcat_overlay.conf . Enablement is pending changes to CI infra which has a lot of dependencies to the current plaintext logs, but we have full capability to employ dictionaries.

Comment (lgirdwood) [reply scripts/xtensa-build-zephyr.py:901]: @andyross I'm good to go ahead without CI being ready - can you wrap any transition a a Kconfig which we can enables as CI catches up.

Comment (andyross) [reply scripts/xtensa-build-zephyr.py:901]: CI should be fine, almost all of it was just waiting for the missing Zephyr patch mentioned above, which is in now that #9278 has merged.[1]

Should have it up today, July-4th-distractions willing.

[1] Also there was a tweak needed for testbench, which builds the SOF code in a third (!) cmake environment to make a shared library for tplg_parser which early-exits (!!) from the legacy CMakeLists.txt, so I needed to move where the generation was included.

**Comment by andyross** [comment]
Rebased.  One more separate-toplevel-cmake spot for sof-logger needed.  Cleaned up some of the checkpatch warnings.  CI looked good otherwise as of the last update.  Removing draft status.

File: src/audio/eq_iir/eq_iir.c
Comment (lyakh) [reply src/audio/eq_iir/eq_iir.c:40]: @andyross is this planned for a future PR?

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Merging due to large surface area of string changes to reduce conflicts.

**Comment by andyross** [comment]
Wow, that was fast.  Will push a cleanup patch to sof-test that removes the final error (which is by-design now), maybe in combination with some more logging cleanup to more firmly wall off the old dma-trace code from Zephyr builds such that they don't need the (ignored) trace-context UUIDs in core at all.

**Comment by lgirdwood** [comment]
> > Wow, that was fast. Will push a cleanup patch to sof-test that removes the final error (which is by-design now), maybe in combination with some more logging cleanup to more firmly wall off the old dma-trace code from Zephyr builds such that they don't need the (ignored) trace-context UUIDs in core at all.
> 
> @andyross please also add support for LLEXT UUIDs

Thanks @andyross - yeah, lets deal with the final error as the next step and follow up with llext and more cleanup.

**Pull request merged**
